geps/gep-1748.md

```diff
@@ -0,0 +1,140 @@
+would like to route to. For example, the following configuration will send
+requests with paths prefixed with “/west” to the store-west ServiceImport, and
+“/east” to the store-east ServiceImport. Requests that don’t match either of
+these paths will be routed to the “store” ServiceImport which represents a
+superset of the “store-west” and “store-east” ServiceImports.
+
+```yaml
+{% include 'standard/multicluster/httproute-location.yaml' %}
+```
+
+#### Advanced Routing With ServiceImports
+
+All Routing capabilities in Gateway API should apply equally whether the backend
+is a Service or ServiceImport. For example, when routing to a system with
+multiple read replicas, it could be beneficial to route requests based on HTTP
+Method. In the following example, requests with POST, PUT, and DELETE methods
+are routed to `api-primary` while the rest are routed to `api-replicas`:
+
+```yaml
+{% include 'standard/multicluster/httproute-method.yaml' %}
+```
+
+#### Routing to Both Services and ServiceImports
+
+There are some situations where it will be useful to split traffic between a
+Service and ServiceImport. In the following example, 90% of traffic would go to
+endpoints attached to the cluster-local "store" Service, and the remaining 10%
+would go to endpoints attached to the Multi-Cluster "store-global" Service:
+
+```yaml
+{% include 'standard/multicluster/httproute-hybrid.yaml' %}
+```
+
+
+#### Cross-Namespace References with ReferenceGrant
+
+It is possible to use ReferenceGrant to enable cross-namespace references to
+ServiceImports. For example, the following HTTPRoute would forward traffic to
+endpoints attached to the “bar” Service in the “bar” namespace:
+
+```yaml
+{% include 'standard/multicluster/httproute-referencegrant.yaml' %}
+```
+
+### Mesh: ServiceImport as Parent
+
+In some cases, you may want to override traffic destined for a Multi-Cluster
+Service with a mesh. As part of the broader GAMMA initiative, ServiceImport can
+be used in the same way that Service is used as a ParentRef. When a Service is
+specified as a parent, meshes will intercept traffic destined for the ClusterIP
+of the Service and apply any policies or routing decisions defined by the Route.
+Similarly, when a ServiceImport is specified as a parent, meshes will intercept
+traffic destined for the ClusterSetIP and apply any policies or routing
+decisions defined by the Route. In the following example, the mesh would
+intercept traffic destined for the store ClusterSetIP and redirect it to
+endpoints for the local Service within the same cluster.
+
+```yaml
+{% include 'standard/multicluster/httproute-gamma.yaml' %}
+```
+
+### Services vs ServiceImports
+
+It is important that all implementations provide a consistent experience. That
+means that references to Services MUST always be interpreted as references to
+endpoints within the same cluster for that Service. References to ServiceImports
```

Comment:
>I wanted to flag that this seems to prohibit "transparent-proxy" mesh functionality that could automatically route traffic to a Service to viable multicluster endpoints unless explicitly encoded in Route configuration.

Yeah, that's the intent here. The idea would be that users should explicitly be able to request multicluster endpoints/routing, but never have that happen automatically. 

>For GAMMA configuration, this could be enabled per Service by writing a Route to direct all traffic intended for a Service to a corresponding ServiceImport backend to spread load across multicluster endpoints (as an alternative to explicitly updating application upstreams from *.cluster.local to *.clusterset.local), which seems reasonable enough, but does have some configuration overhead for enabling clusterset-wide routing for all multicluster services.

Agree that that's a potential workaround. I have updated this to a `SHOULD` for now based on @keithmattix's comment above, but would like to see Gateway API/GAMMA usage adopt the Kubernetes pattern that multicluster routing is something you opt in to. Maybe the more important concept would be that exporting a Service in one cluster is required for it to be accessible in other clusters. Open to input here, but if that's a workable solution, it would definitely be more consistent with the rest of Kubernetes.

I am interested in finding ways to lower the implementation overhead for MCS today, such as the current proposal that would remove the requirement to propagate EndpointSlices to each cluster. I'm hoping this can become just a configuration thing where users simply have to opt-in to exporting a Service and the rest is lightweight and automatic.

Reply:
Resolving this for now, let me know if I missed anything @mikemorris

Reply:
> The idea would be that users should explicitly be able to request multicluster endpoints/routing, but never have that happen automatically.

Was talking with @howardjohn a bit about this yesterday, and the concern is that this approach can force multicluster load balancing or failover to be a concern of the service _consumer_ rather than the _producer_ who may best understand how to shape incoming traffic to a service under their management to ensure reliability and high availability.

Consumer opt-in can make sense for service discovery use cases when DNS is the primary interface to return workload addresses to dial directly (Consul's service discovery uses [prepared queries](https://developer.hashicorp.com/consul/docs/services/discovery/dns-dynamic-lookups) for similar client-defined routing behavior), but allowing producer-defined routing rules to direct `.cluster.local` requests to off-cluster endpoints is an important mesh capability.

I think we have enough wiggle room with the change to `SHOULD`, but definitely want to get this explored in more depth as a GAMMA+MCS user stories.

> would like to see Gateway API/GAMMA usage adopt the Kubernetes pattern that multicluster routing is something you opt in to

For me the key is which persona is opting in to this functionality. Agreed it would be dangerous for a cluster/mesh administrator to enable global multi-cluster routing (it's nice for demos though 😉), but we shouldn't have to rely on all downstream consumers of a service understanding when and how to configure multi-cluster routing for each upstream service they depend on.

> Maybe the more important concept would be that exporting a Service in one cluster is required for it to be accessible in other clusters.

Is this not already an assumption/requirement? **Definitely agreed** with this - meshes _may_ offer a higher-level API to do something like automatically create ServiceExports for every service in a given namespace, but having services explicitly exported in order for an MCS implementation to create corresponding ServiceImports on other clusters in the group/clusterset is definitely essential - some services should likely always stay cluster-local.

> I am interested in finding ways to lower the implementation overhead for MCS today

Same! I'm excited to see progress on conformance tests, and getting more clarity around some of these behaviors or edge cases is going to be important for MCS to be a useful building block for multi-cluster service mesh deployments.